### ISSUE TITLE
docker compose build fails with python 3.12 (but not with python 3.11)

### ISSUE THREAD

**Issue by fniessink** [Pythagora-io/gpt-pilot#126]
Building the current main on macOS Sonoma 14.0 (M1) fails with the output below. Changing the first line of the Dockerfile to `FROM python:3.11` fixes the issue.

Relevant output from `docker compose build` with current main:

```console
7.930 Building wheels for collected packages: peewee, psycopg2-binary, regex, tiktoken
7.931   Building wheel for peewee (pyproject.toml): started
8.651   Building wheel for peewee (pyproject.toml): finished with status 'done'
8.652   Created wheel for peewee: filename=peewee-3.16.2-py3-none-any.whl size=135166 sha256=1d52c54dea2437a097ff1d27ef41db0935427c54db10838ba47fd1ae1a02cfe5
8.652   Stored in directory: /tmp/pip-ephem-wheel-cache-v6e9eiik/wheels/ef/4f/25/82c0400740e01c9e720d0eb6dd297dcba32be5180fa5f65782
8.655   Building wheel for psycopg2-binary (setup.py): started
13.19   Building wheel for psycopg2-binary (setup.py): finished with status 'done'
13.20   Created wheel for psycopg2-binary: filename=psycopg2_binary-2.9.6-cp312-cp312-linux_aarch64.whl size=512651 sha256=af270df9f3c1df91bfe18b5bde45198e878e6a13d1c8195afab66b85952ccc39
13.20   Stored in directory: /tmp/pip-ephem-wheel-cache-v6e9eiik/wheels/b9/f5/6f/ea1a079f89119021dc6310b512ae3bc14956d09826f02eaf38
13.20   Building wheel for regex (pyproject.toml): started
20.43   Building wheel for regex (pyproject.toml): finished with status 'done'
20.43   Created wheel for regex: filename=regex-2023.6.3-cp312-cp312-linux_aarch64.whl size=769725 sha256=d58303ad9bc0ec1da8555ec9a0213dd952f4bb9cf4f5808f678a91f65d3623c7
20.43   Stored in directory: /tmp/pip-ephem-wheel-cache-v6e9eiik/wheels/62/21/26/24e73e5cb8863075e1ce495d0ae12d6c6b89f81844efb9e586
20.44   Building wheel for tiktoken (pyproject.toml): started
20.59   Building wheel for tiktoken (pyproject.toml): finished with status 'error'
20.60   error: subprocess-exited-with-error
20.60
20.60   × Building wheel for tiktoken (pyproject.toml) did not run successfully.
20.60   │ exit code: 1
20.60   ╰─> [37 lines of output]
20.60       running bdist_wheel
20.60       running build
20.60       running build_py
20.60       creating build
20.60       creating build/lib.linux-aarch64-cpython-312
20.60       creating build/lib.linux-aarch64-cpython-312/tiktoken
20.60       copying tiktoken/load.py -> build/lib.linux-aarch64-cpython-312/tiktoken
20.60       copying tiktoken/__init__.py -> build/lib.linux-aarch64-cpython-312/tiktoken
20.60       copying tiktoken/registry.py -> build/lib.linux-aarch64-cpython-312/tiktoken
20.60       copying tiktoken/core.py -> build/lib.linux-aarch64-cpython-312/tiktoken
20.60       copying tiktoken/model.py -> build/lib.linux-aarch64-cpython-312/tiktoken
20.60       creating build/lib.linux-aarch64-cpython-312/tiktoken_ext
20.60       copying tiktoken_ext/openai_public.py -> build/lib.linux-aarch64-cpython-312/tiktoken_ext
20.60       running egg_info
20.60       writing tiktoken.egg-info/PKG-INFO
20.60       writing dependency_links to tiktoken.egg-info/dependency_links.txt
20.60       writing requirements to tiktoken.egg-info/requires.txt
20.60       writing top-level names to tiktoken.egg-info/top_level.txt
20.60       reading manifest file 'tiktoken.egg-info/SOURCES.txt'
20.60       reading manifest template 'MANIFEST.in'
20.60       warning: no files found matching 'Makefile'
20.60       adding license file 'LICENSE'
20.60       writing manifest file 'tiktoken.egg-info/SOURCES.txt'
20.60       copying tiktoken/py.typed -> build/lib.linux-aarch64-cpython-312/tiktoken
20.60       running build_ext
20.60       running build_rust
20.60       error: can't find Rust compiler
20.60
20.60       If you are using an outdated pip version, it is possible a prebuilt wheel is available for this package but pip is not able to install from it. Installing from the wheel would avoid the need for a Rust compiler.
20.60
20.60       To update pip, run:
20.60
20.60           pip install --upgrade pip
20.60
20.60       and then retry package installation.
20.60
20.60       If you did intend to build this package from source, try installing a Rust compiler from your system package manager and ensure it is on the PATH during installation. Alternatively, rustup (available at https://rustup.rs) is the recommended way to download and update the Rust compiler toolchain.
20.60       [end of output]
20.60
20.60   note: This error originates from a subprocess, and is likely not a problem with pip.
20.60   ERROR: Failed building wheel for tiktoken
20.60 Successfully built peewee psycopg2-binary regex
20.60 Failed to build tiktoken
20.60 ERROR: Could not build wheels for tiktoken, which is required to install pyproject.toml-based projects
------
failed to solve: process "/bin/sh -c pip install --no-cache-dir -r requirements.txt" did not complete successfully: exit code: 1
```

**Comment by Kwallie010** [comment]
Thank you for providing the fix at least!

**Comment by iddar** [comment]
> Building the current main on macOS Sonoma 14.0 (M1) fails with the output below. Changing the first line of the Dockerfile to `FROM python:3.11` fixes the issue.
> 


@fniessink  thanks for the quick solution

**Comment by nalbion** [comment]
Thanks @fniessink this will be resolved in PR #125
